### PR TITLE
Add gitpod integration

### PR DESCRIPTION
**Pull request by timja**
Gitpod is a cloud development environment that you can use to quickly get up to speed on a project by just clicking 'Open in gitpod'.

It can pre-build workspaces so you don't have to wait for it to get setup.

You can see it in use on https://github.com/jenkinsci/design-library-plugin/pull/67

Supported IDEs are VS Code and Intelli J.

What do you think?

cc @Nancy-Chauhan

(for the below button to appear automatically I would need to install the GitHub app on this repository which I can do after this is accepted)

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6590"></a>

### Proposed changelog entries

* Entry 1: Issue, Human-readable Text
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

### CODE REVIEW

File: .gitpod.yml
Comment (akosyakov) [inline .gitpod.yml:4]: You may consider to introduce a custom Dockerfile instead [[1](https://www.gitpod.io/docs/config-docker#configure-a-custom-dockerfile)]

File: .gitpod.yml
Comment (akosyakov) [inline .gitpod.yml:15]: I wonder whether this line is really needed? I was able to get UI without it too and redirection did not work anyway.

How `GIT_POD_URL` is populated in this file?

Comment (timja) [reply .gitpod.yml:15]: let me try without it, it helped before I found out about the `host=0.0.0.0` thing.

File: .gitpod.yml
Comment (timja) [inline .gitpod.yml:16]:
```suggestion
      mvn jetty:run -Dhost=0.0.0.0
```

File: .gitpod.yml
Comment (timja) [reply .gitpod.yml:15]: Ah I had forgotten to commit the host thing.

Redirects should work now.

If I don't add this file then you get this warning:

image

Which means certain things wouldn't work but generally it's fine.

So could omit it but probably doesn't do any harm.

----

an alternative may be to install the configuration as code plugin in the init script which would allow more config to be done if people wanted.

----

> How `GIT_POD_URL` is populated in this file?

by running `gp url 8080`

File: .gitpod.yml
Comment (timja) [reply .gitpod.yml:4]: Does it make much of a difference? It will still have to build the docker image anyway.

I tried this last week sometime with the dockerfile and it didn't work right for me, but I probably missed something.

https://github.com/timja/jenkins/blob/b277277379722a1c8bdbdc61f675817c2bb89f27/.gitpod.yml#L17-L18

**Comment by timja** [comment]
(misclick closing it)

**Comment by timja** [comment]
One thing that can be annoying is that most of the time starting up is building maven dependencies.

If people adopt this I was wondering about publishing all our maven deps as a tarball on each release and then just fetching them on startup / in the docker file.

downloading a single tar ball from GitHub CDN be a lot quicker than fetching each artifact individually from artifactory I think

File: .gitpod.yml
Comment (akosyakov) [reply .gitpod.yml:4]: It is a bit different from point of caching. Gitpod has 2 builds one for changes in Dockerfile and other in prebuild for changes in repo. By moving tools setup which is not dependent on code in repo to Dockerfile you decrease time of prebuilds for all branches.

**Comment by akosyakov** [comment]
> One thing that can be annoying is that most of the time starting up is building maven dependencies.

If you enable [prebuiilds](https://www.gitpod.io/docs/prebuilds#prebuilds) then they be built ahead of times.

File: .gitpod.yml
Comment (akosyakov) [reply .gitpod.yml:4]: It should be like:
```yaml
image:
  file: .gitpod.Dockerfile
```

**Comment by timja** [comment]
> > One thing that can be annoying is that most of the time starting up is building maven dependencies.
> 
> If you enable [prebuiilds](https://www.gitpod.io/docs/prebuilds#prebuilds) then they be built ahead of times.

it is enabled on the design library plugin and it seems to occasionally use it but generally not.
(not that i've tested it out that much).

It takes something like 6 seconds to start design library if dependencies are there, or around 5 minutes if not.

**Comment by akosyakov** [comment]
@timja Do you have a [project](https://www.gitpod.io/docs/teams-and-projects#projects) already? You should be able to see when your repo there and all associated prebuilds with a state whether they failed or not. If they are successful then it should always be available. Usually it will take 5 mins either because it failed or you pressed to skip a prebuild.

**Comment by timja** [comment]
Maybe it's working,
https://gitpod.io/projects/design-library-plugin/861e3af9-da30-4f53-a508-f62c91dacd2a

says it took 16 seconds.

**Comment by akosyakov** [comment]
You need to enable indexing for IntelliJ in prebuilds too that you don't wait for indexing, resolving deps and so on: https://www.gitpod.io/docs/ides-and-editors/intellij#indexing-using-prebuilds

**Comment by akosyakov** [comment]
But otherwise I see in output:
```
🤙 This task ran as a workspace prebuild
🎉 Well done on saving 3 minutes
```

So it was prebuilt.

**Comment by timja** [comment]
> You need to enable indexing for IntelliJ in prebuilds too that you don't wait for indexing, resolving deps and so on: [gitpod.io/docs/ides-and-editors/intellij#indexing-using-prebuilds](https://www.gitpod.io/docs/ides-and-editors/intellij#indexing-using-prebuilds)

ah ok, I noticed that and it was a bit annoying.
I've added the recommendation although the docs don't seem to say what stable vs latest do here (unless I missed it).

File: .gitpod.yml
Comment (felladrin) [inline .gitpod.yml:18]: Hi, @timja!

The issue with "gp preview" not working in JetBrains IDEs has been fixed and deployed (https://github.com/gitpod-io/gitpod/issues/10118), so you can remove `gp await-port 63342` from from here:

```suggestion
  - command: gp await-port 8080 && gp url 8080 && gp preview $(gp url 8080)/jenkins/
```

It can also be removed from [jenkinsci/design-library-plugin's .gitpod.yml](https://github.com/jenkinsci/design-library-plugin/blob/4b1d2dd3da7ecbf61e4e3a466575a23dddad2adc/.gitpod.yml#L8).

Comment (timja) [reply .gitpod.yml]: thanks!

File: CONTRIBUTING.md
Comment (NotMyFault) [inline CONTRIBUTING.md:69]:
```suggestion
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#https://github.com/jenkinsci/jenkins)
```
👀

Comment (NotMyFault) [reply CONTRIBUTING.md]: Actually, doesn't that open the upstream repository? You'd need to open your forked repository to propose changes.

Comment (timja) [reply CONTRIBUTING.md]: No that would start you from some super outdated fork probably.

The normal workflow is to start from upstream/master, then create branch off of that, push to your fork and create a PR from it.

So when you get there in gitpod using the above button you will just need to run `git switch -c my-branch`

Comment (NotMyFault) [reply CONTRIBUTING.md]: Oic, thanks for the clarification.

File: .gitpod.yml
Comment (timja) [reply .gitpod.yml]: that worked, I think I was missing setting maven to be the default version before

**Review by NotMyFault — APPROVED** [review]

**Comment by Nancy-Chauhan** [comment]
Hey 👋 @timja is this PR good to go 🚀  ?

**Review by jtnord — APPROVED** [review]
I do not see reason why we should not do this.

The maintenance burden looks low, and if this helps get more/better contributions then 🥳

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: .gitpod.yml
Comment (basil) [inline .gitpod.yml:6]: Pulling down a shell script from a user's personal gist seems like a workaround to me. Can we not e.g. set a system property when invoking `mvn jetty:run` to customize this value? Seems like the ideal solution would be to make this customizable by a system property if it isn't already and then to pass that property to `mvn jetty:run`. This would present a (semi) supported interface for setting this value, whereas manually writing out an XML file is an awkward workaround and could become out of date in the future if the object model changes. If I am missing something and implementing a system property for this is either infeasible or a large amount of work, then I could possibly be convinced otherwise about the system property approach, but I doubt you will be able to convince me that putting a link to `https://gist.githubusercontent.com/timja` in Jenkins core is a good long-term solution. :smile:

Comment (timja) [inline .gitpod.yml:6]:
```suggestion
```

Comment (timja) [reply .gitpod.yml]: not going to debate this now, there's no interface / property I'm aware of and not inclined to do it at this time.
I've removed setting the property and users will just get prompted to save it once, (the same experience they currently get with `jetty:run` locally)

File: .gitpod/Dockerfile
Comment (jglick) [inline .gitpod/Dockerfile:4]: see also #6491

File: CONTRIBUTING.md
Comment (jglick) [inline CONTRIBUTING.md:69]: IIUC https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6590 works already?

File: .gitpod.yml
Comment (jglick) [inline .gitpod.yml:3]: Could add `-ntp` here but I guess it is valuable to see what is taking so long.

File: .gitpod.yml
Comment (jglick) [inline .gitpod.yml:6]: or
```suggestion
      mvn -pl war jetty:run -Dhost=0.0.0.0
```

**Review by jglick — COMMENTED** [review]
Run fails for me:

```
SEVERE  jenkins.InitReactorRunner$1#onTaskFailed: Failed Contributed.load
java.lang.NullPointerException
        at jenkins.security.CustomClassFilter$Contributed.load(CustomClassFilter.java:155)
```

https://github.com/jenkinsci/jenkins/blob/b3d64bc71ddc17c81c9c40bf0a4cf0f0f04ba6e7/core/src/main/java/jenkins/security/CustomClassFilter.java#L155 suggests a problem with `@Extension` indexing. I think you may need to do a clean build (which fixes the problem for me); maybe a clash with background compilation by the IDE?

File: .gitpod.yml
Comment (timja) [reply .gitpod.yml:3]: Yeah have thought about it but I think it’s better to see in this case

File: CONTRIBUTING.md
Comment (timja) [reply CONTRIBUTING.md:69]: Works fine last I checked

**Comment by timja** [comment]
Using which url fails for you? And doing what?

**Comment by jglick** [comment]
> Using which url fails for you?

https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6590

> And doing what?

Just opening the IDE and I guess installing the Java extension when prompted to do so. `jetty:run` tries to start Jenkins, which crashes with fatal boot errors, apparently due to missing extensions.

**Comment by timja** [comment]
> Just opening the IDE and I guess installing the Java extension when prompted to do so

Using intellij or vs code? (I haven't tried vs code out with this although intellij is working fine for me)

**Comment by jglick** [comment]
> Using intellij or vs code?

VS Code is what GitPod offers you. (Well, you still have the option to pick Theia but it discourages you from doing so.)

**Comment by timja** [comment]
> > Using intellij or vs code?
> 
> VS Code is what GitPod offers you. (Well, you still have the option to pick Theia but it discourages you from doing so.)

I can reproduce this if you use VS Code *and* install the Java language pack.
If you don't install the language pack it works fine.

IntelliJ GitPod integration works fine too and is what (I assume) more people would use.

Do you think this is fine for an initial version?

**Comment by akosyakov** [comment]
You can add to .gitpod.yml
```yml
vscode:
  extensions:
    - vscjava.vscode-java-pack
```

**Comment by akosyakov** [comment]
> VS Code is what GitPod offers you. (Well, you still have the option to pick Theia but it discourages you from doing so.)

Gitpod does not offer Theia for one year already 🤔 When you first time login in Gitpod it will ask you whether you would like to use VS Code, IntelliJ or other editors.

**Comment by timja** [comment]
> You can add to .gitpod.yml
> 
> ```yaml
> vscode:
>   extensions:
>     - vscjava.vscode-java-pack
> ```

I don't know if we want that though, adding the java language pack causes Jenkins to not start in the dev server

File: .gitpod.yml
Comment (jglick) [inline .gitpod.yml:3]: I wonder if
```suggestion
      mvn -am -pl war,bom -P quick-build clean install
```
works around the extension index problem with the IDE. As noted in https://github.com/jenkinsci/jenkins/pull/6590#pullrequestreview-992504596 a clean build _after_ failure works.

**Comment by jglick** [comment]
https://www.gitpod.io/docs/ides-and-editors/configure-your-editor-ide#configuring-an-ide-preference says you can configure an IDE per project but https://www.gitpod.io/docs/references/gitpod-yml does not seem to have an option to _select_ an IDE.

File: .gitpod.yml
Comment (jglick) [inline .gitpod.yml:24]: prompt
despite this

Comment (timja) [reply .gitpod.yml:24]: not for me, but I assume resolved given follow up comments?

**Comment by timja** [comment]
> Seems to work now! Note that the internal browser is not completely functional; e.g. after creating a project you are not redirected to the config screen.

Yep I just hit the same thing when testing it out, weirdly it creates a job but fails to redirect you and if you click again it says it exists.

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: .gitpod/Dockerfile
Comment (timja) [reply .gitpod/Dockerfile:4]: Yep that's the recommended way of handling this, and `mvn` wrapper would be used by default from what I remember

File: .gitpod.yml
Comment (jglick) [reply .gitpod.yml:24]: > I assume resolved

No, I just clicked **Install** and proceeded.

File: CONTRIBUTING.md
Comment (jglick) [inline CONTRIBUTING.md:67]:
```suggestion
You can open this project as a [Gitpod workspace](https://www.gitpod.io/) which comes pre-configured with all the tools you will need.
You can use IntelliJ IDEA (preferred) or VS Code (alternate) in the browser.
```

File: CONTRIBUTING.md
Comment (jglick) [inline CONTRIBUTING.md:71]:
```suggestion
If you prefer using IntelliJ IDEA (desktop), you can setup Gitpod integration with JetBrains Gateway using the instructions on [gitpod.io](https://www.gitpod.io/docs/ides-and-editors/intellij),
```
?

**Review by jglick — APPROVED** [review]

File: CONTRIBUTING.md
Comment (timja) [reply CONTRIBUTING.md:72]: Unneeded I think

**Pull request merged**
